src/main/java/sparksoniq/spark/iterator/flowr/ForClauseSparkIterator.java

```diff
@@ -20,45 +20,171 @@
+        }
+
+        // execution reaches here when there are no more results
+        _child.close();
+        this._hasNext = false;
+    }
+
+    /**
+     * _expression has to be open prior to call.
+     * @return true if _nextLocalTupleResult is set and _hasNext is true, false otherwise
+     */
+    private boolean setResultFromExpression() {
+        if (_expression.hasNext()) {     // if expression returns a value, set it as next
+            List<Item> results = new ArrayList<>();
+            results.add(_expression.next());
+            FlworTuple newTuple = new FlworTuple(_inputTuple, _variableName, results);
+            _nextLocalTupleResult = newTuple;
+            this._hasNext = true;
+            return true;
+        } else {
+            _expression.close();
+            return false;
+        }
+    }
+
+    private void initialFor_setNextLocalTupleResult() {
+        if (_expression.hasNext()) {
+            List<Item> results = new ArrayList<>();
+            results.add(_expression.next());
+            FlworTuple newTuple = new FlworTuple(_variableName, results);
+            _nextLocalTupleResult = newTuple;
+            this._hasNext = true;
+        } else {
+            _expression.close();
+            this._hasNext = false;
+        }
+    }
+
+    @Override
+    public void open(DynamicContext context) {
+        super.open(context);
+
+        // isRDD checks omitted, as open is used for non-RDD(local) operations
+
+        if (this._child != null) { //if it's not a start clause
+            _child.open(_currentDynamicContext);
+            _tupleContext = new DynamicContext(_currentDynamicContext);     // assign current context as parent
+
+            setNextLocalTupleResult();
+
+        } else {    //if it's a start clause
+            _expression.open(this._currentDynamicContext);
+
+            initialFor_setNextLocalTupleResult();
+        }
+    }
+
+    @Override
+    public void close() {
+        if (_child != null) {
+            _child.close();
+        }
     }
 
 
     @Override
     public JavaRDD<FlworTuple> getRDD() {
-        if (this._rdd == null) {
-            String variableReference = ((VariableReferenceIterator)this._children.get(0)).getVariableName();
-            RuntimeIterator assignmentExpression = this._children.get(1);
+        if (this._rdd == null) {        //if it's a start clause
             JavaRDD<Item> initialRdd = null;
-            //if it's a start clause
+
             if (this._child == null) {
-                initialRdd = this.getNewRDDFromExpression(assignmentExpression);
-                this._rdd = initialRdd.map(new InitialForClauseClosure(variableReference));
-            } else {
-            //if it's not a start clause
-                this._rdd = this._child.getRDD();
-                this._rdd = this._rdd.flatMap(new ForClauseClosure(assignmentExpression, variableReference));
+                initialRdd = this.getNewRDDFromExpression(_expression);
```

Comment:
I think that can even be simplified, because the else case in getNewRDDFromExpression will never happen (it is now handled with the newly implemented local execution). So you can directly ask for the expression RDD here and remove that function.

Reply:
Yes that's a very good suggestion. However, we have to postpone this until all isRDD methods are overridden in all flwor clauses. Otherwise it causes Iterator is not an RDD exceptions for LocalIterators in the existing test cases. I'll open an issue for later

Reply:
Yes, an issue is fine. It's good practice to keep PRs small.